pandas/core/reshape/merge.py

```diff
@@ -915,11 +916,15 @@ def _maybe_coerce_merge_keys(self):
             # Houston, we have a problem!
             # let's coerce to object
             if name in self.left.columns:
+                cat = is_categorical_dtype(lk)
+                typ = lk.categories.dtype if cat else object
```

Comment:
Not in that elif block, but it could be done in the one above:

```python
            if is_categorical_dtype(lk) and is_categorical_dtype(rk):
                if lk.is_dtype_equal(rk):
                    continue

                lk_to = lk.categories.dtype
                rk_to = rk.categories.dtype
```
but that doesn't seem cleaner to me - if we spread the `lk_to`/`rk_to` logic all over the method, then we make it much more difficult to debug compared with having all of the coercion logic in the block at the bottom where the coercion happens.